src/main/java/com/icthh/xm/ms/entity/service/XmEntitySpecService.java

```diff
@@ -430,48 +364,71 @@ private boolean isObject(JsonNode schemaNode) {
         return getNodeType(schemaNode).equals(OBJECT) && schemaNode.has("properties");
     }
 
-    @Override
-    @SneakyThrows
-    @IgnoreLogginAspect
-    public void onRefresh(String updatedKey, String config) {
-        String specificationPathPattern = applicationProperties.getSpecificationPathPattern();
-        try {
-            String tenant = matcher.extractUriTemplateVariables(specificationPathPattern, updatedKey).get(TENANT_NAME);
-            if (org.apache.commons.lang3.StringUtils.isBlank(config)) {
-                types.remove(tenant);
-                return;
-            }
-            XmEntitySpec spec = mapper.readValue(config, XmEntitySpec.class);
-            Map<String, TypeSpec> value = toTypeSpecsMap(spec);
-            types.put(tenant, value);
-            entityCustomPrivilegeService.updateApplicationPermission(value, tenant);
-            log.info("Specification was for tenant {} updated", tenant);
-        } catch (Exception e) {
-            log.error("Error read xm specification from path " + updatedKey, e);
-        }
+    private String getTenantKeyValue() {
+        return TenantContextUtils.getRequiredTenantKeyValue(tenantContextHolder);
     }
 
-    @Override
-    public boolean isListeningConfiguration(String updatedKey) {
-        String specificationPathPattern = applicationProperties.getSpecificationPathPattern();
-        return matcher.match(specificationPathPattern, updatedKey);
+    private Map<String, TypeSpec> toTypeSpecsMap(XmEntitySpec xmEntitySpec) {
+        List<TypeSpec> typeSpecs = xmEntitySpec.getTypes();
+        if (isEmpty(typeSpecs)) {
+            return Collections.emptyMap();
+        } else {
+            // Convert List<TypeSpec> to Map<key, TypeSpec>
+            Map<String, TypeSpec> result = typeSpecs.stream()
+                .collect(Collectors.toMap(TypeSpec::getKey, Function.identity(),
+                    (u, v) -> {
+                        throw new IllegalStateException(String.format("Duplicate key %s", u));
+                    }, LinkedHashMap::new));
+
+            // add inheritance
+            inheritance(result);
+            processUniqueFields(result);
+
+            return result;
+        }
     }
 
-    @Override
-    public void onInit(String key, String config) {
-        if (isListeningConfiguration(key)) {
-            onRefresh(key, config);
+    /**
+     * XM Entity Type specifications support inheritance based on key hierarchy
+     * with data extension.
+     */
+    private void inheritance(Map<String, TypeSpec> types) {
```

Comment:
i spend time and try to compare this methods. and it same?
just reodreding or what happend?

Reply:
yes, only reordering